### PR TITLE
Add support for registering an analyzer for concurrent execution of its actions

### LINKED ISSUE #6737: Expose metadata to mark analyzers for concurrent execution
[NOTE: Split from https://github.com/dotnet/roslyn/issues/1565, which now only tracks *enabling incremental analysis for 3rd party analyzers in IDE*]

This issue tracks exposing metadata from analyzers to mark them as thread-safe, such that analyzer driver will be free to make concurrent action callbacks into the analyzer instance.

### PR DESCRIPTION
**Pull request by mavasani**
By default, analyzer driver never makes concurrent callbacks into a single analyzer instance. This change adds a new public API `AnalysisContext.RegisterConcurrentExecution` to enable thread-safe analyzers to opt into receiving concurrent callbacks, which should theoretically improve performance of such analyzers for concurrent builds.

Fixes #6737

### CODE REVIEW

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -76,6 +76,7 @@ virtual Microsoft.CodeAnalysis.MetadataReferenceResolver.ResolveMissingAssembly(
 virtual Microsoft.CodeAnalysis.SourceReferenceResolver.ReadText(string resolvedPath) -> Microsoft.CodeAnalysis.Text.SourceText
 Microsoft.CodeAnalysis.SemanticModel.GetOperation(Microsoft.CodeAnalysis.SyntaxNode node, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> Microsoft.CodeAnalysis.Semantics.IOperation
 abstract Microsoft.CodeAnalysis.SemanticModel.GetOperationCore(Microsoft.CodeAnalysis.SyntaxNode node, System.Threading.CancellationToken cancellationToken) -> Microsoft.CodeAnalysis.Semantics.IOperation
+abstract Microsoft.CodeAnalysis.Diagnostics.AnalysisContext.RegisterConcurrentExecution() -> void
```
Comment (mavasani) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: This is the public API being added.

**Comment by mavasani** [comment]
@JohnHamby @gafter @srivatsn @heejaechang can you please review?

**Comment by JohnHamby** [comment]
In a quick scan of the code, I didn't see anything preventing concurrent execution of actions that are too closely related. For example, it's probably not sound to simultaneously execute two syntax node actions registered by a code block start action for a given code block (because the two syntax node actions are very likely to be sharing state for the code block). Requiring an analyzer that does this to have no concurrency is too restrictive.

At some point in the past I wrote up a proposed set of rules for when two actions are too closely related to execute in parallel.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs [hunk 2]
```diff
@@ -162,6 +162,18 @@ public void RegisterOperationAction(Action<OperationAnalysisContext> action, par
...
+        /// Register for concurrent execution of analyzer actions registered by this analyzer.
+        /// An analyzer that registers for concurrent execution should be more performant then a non-concurrent analyzer.
+        /// However, such analyzers must ensure that its actions are implemented in a thread-safe manner.
+        /// </summary>
+        /// <remarks>
+        /// Even when an analyzer registers for concurrent execution, certain related actions are *never* executed concurrently.
+        /// For example, end actions registered on any analysis unit (compilation, code block, operation block, etc.) are by definition semantically dependent on analysis from non-end actions registered on the same analysis unit.
+        /// Hence, end actions are never executed concurrently with non-end actions operating on the same analysis unit.
+        /// </remarks>
+        public abstract void RegisterConcurrentExecution();
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: is it decide so that people can't flip it more than once?

**Comment by heejaechang** [comment]
:+1:

**Comment by mavasani** [comment]
> it's probably not sound to simultaneously execute two syntax node actions registered by a code block start action for a given code block (because the two syntax node actions are very likely to be sharing state for the code block). 

I am unsure why we would add such a restriction. What if these actions themselves are thread-safe and can deal with access/update of shared state concurrently? Adding such a restriction will prevent further performance improvements for such an analyzer.

I think we are talking about 3 layers here:
1. Completely non-concurrent
2. Maximally concurrent and thread-safe
3. Partially concurrent and thread safe

I propose we treat the new API to enforce analyzer needs to be (2) - where only semantically dependent actions are not executed concurrently. End actions by definition depend on non-end actions executing on same analysis unit (compilation, code block, operation block, etc.) and hence cannot be executed before them - driver guarantees that. However, sort of related actions that you mention can be written in thread-safe way and executed concurrently without any functional change in the output.
If and when we get to adding support for 3., we can allow authors to define subset of actions or action kinds that it deems thread safe.

**Comment by heejaechang** [comment]
since we never promised ordering between same level actions (syntax tree, syntax node, code block), I think running those concurrently doesnt break any contracts.

we never promised we will run syntax node actions in in-order tree traversal or pre or post, or tree in the order it shows up in compilation and etc.

so, I think it is fine running them concurrently and let the author deal with managing concurrency.

**Comment by heejaechang** [comment]
by the way, one thing, I think we need to explicitly state that even if one is registered to be executed concurrently, Host can still decide it runs it non-concurrent manner. basically, the call is to let host know characteristic of the analyzer, not to control behavior of the host.

host behavior is completely owned by the host (driver) and analyzer provider can't change it in any way.

**Comment by JohnHamby** [comment]
I hunted up the most succinct statement of the "related actions" rules I could find, and pasted it below. The incentive for such restrictions is to make a very high percentage of analyzers achieve maximal safe concurrency without an author having to take locks. Getting locking correct is very difficult, and we shouldn't push the problem entirely onto analyzer authors. The ideal experience is that an author turns on concurrency and an analyzer keeps working.

Here's the last draft of the rules. What we're doing with shared state will necessitate some refinement:

1)      The Initialize methods of two or more different analyzers can run concurrently. We make no guarantees about state shared between analyzers, or about static state.
2)      The actions of two or more different analyzers can run concurrently. We make no guarantees about state shared between analyzers.
3)      Any number of actions that were registered directly by a single analyzer’s Initialize method can run concurrently. We make no guarantees about static state, or about cross-compilation state.
4)      An action A is an ancestor of action B if B was registered either by A or by another action that has A as an ancestor. Two actions X and Y are related if any of these hold:
    a.       X is an ancestor of Y.
    b.      Y is an ancestor of X.
    c.       X and Y have a common ancestor.
    d.      X and Y are the same action, and that action has an ancestor.
5) Any number of actions of a single analyzer can run concurrently so long as no two of those actions are related.

**Comment by heejaechang** [comment]
@JohnHamby I dont think user can understand that rule. so either user will not mark it as concurrently or use lock regardless.

**Comment by mavasani** [comment]
> Getting locking correct is very difficult, and we shouldn't push the problem entirely onto analyzer authors. The ideal experience is that an author turns on concurrency and an analyzer keeps working.

I think analyzer author signing up for concurrency *have* to ensure their actions are thread safe (via locking, concurrent data structures, etc.). They cannot just turn on concurrency and expect their existing analyzer which shares state between actions to keep working.

I agree with your definition of related actions, but this won't be easy for any analyzer author to understand. We shouldn't require them to figure out what actions are related or unrelated and then go and implement thread safety. 

Additionally, one big drawback of this restriction would be that all actions registered within a compilation start action would be related actions and hence wouldn't be executed concurrently - I believe majority of analyzers are written as compilation start analyzers just to fetch some well known types from the compilations, but have no related actions or state sharing.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs [hunk 3]
```diff
@@ -162,6 +162,18 @@ public void RegisterOperationAction(Action<OperationAnalysisContext> action, par
         /// <param name="action">Action to be executed at completion of semantic analysis of an <see cref="IOperation"/>.</param>
         /// <param name="operationKinds">Action will be executed only if an <see cref="IOperation"/>'s Kind matches one of the operation kind values.</param>
         public abstract void RegisterOperationAction(Action<OperationAnalysisContext> action, ImmutableArray<OperationKind> operationKinds);
+
+        /// <summary>
+        /// Register for concurrent execution of analyzer actions registered by this analyzer.
+        /// An analyzer that registers for concurrent execution should be more performant then a non-concurrent analyzer.
```
Comment (JohnHamby) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: How about "can have better performance than" or "can be more efficient than"? I still haven't accepted "performant" as word--the first definition of it I could find online included "Commonly used by Microsofties but not a real word."

At least, "then" should become "than".

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs [hunk 4]
```diff
@@ -162,6 +162,18 @@ public void RegisterOperationAction(Action<OperationAnalysisContext> action, par
...
+        /// However, such analyzers must ensure that its actions are implemented in a thread-safe manner.
```
Comment (JohnHamby) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Maybe "Such an analyzer must ensure that its actions can execute correctly in parallel"? I've heard enough arguments about the precise meaning of "thread safe" that I avoid the term.

At least, "analyzers" and "its" must be reconciled.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 2 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: All of the other methods named "Register***" register an action. Maybe "EnableConcurrency" or "EnableConcurrentActions" or "EnableConcurrentExecution"?

@heejaechang's question is interesting. I don't think we need to support flipping concurrence on and off, but if someone can make a case for it we should discuss that.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs [hunk 5]
```diff
@@ -1133,5 +1133,75 @@ class D
...
+            // Verify no diagnostics.
+            var compilation = GetCompilationWithConcurrentBuildEnabled(source);
+            compilation.VerifyDiagnostics();
+            compilation.VerifyAnalyzerDiagnostics(analyzers);
+        }
+
+        [Fact, WorkItem(6737, "https://github.com/dotnet/roslyn/issues/6737")]
+        public void TestConcurrentAnalyzer()
+        {
+            if (Environment.ProcessorCount <= 1)
```
Comment (Pilchie) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs]: I wouldn't do this.  There's still going to be *some* concurrency around time quantums, etc.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs [hunk 6]
```diff
@@ -1133,5 +1133,75 @@ class D
...
+            }
+
+            var source = string.Empty;
+            var typeCount = 100;
+            var typeNames = new string[typeCount];
+            for (int i = 1; i <= typeCount; i++)
+            {
+                var typeName = $"C{i}";
+                typeNames[i - 1] = typeName;
+                source = source + $"\r\nclass {typeName} {{ }}";
```
Comment (Pilchie) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs]: Ugh.  string concat in a loop.  Make it a `StringBuilder`

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 7]
```diff
@@ -530,7 +546,7 @@ private static ImmutableArray<Diagnostic> FilterDiagnosticsSuppressedInSource(Im
                     var workerTasks = new Task<CompilationCompletedEvent>[workerCount];
                     for (int i = 0; i < workerCount; i++)
                     {
-                        workerTasks[i] = ProcessCompilationEventsCoreAsync(analysisScope, analysisStateOpt, prePopulatedEventQueue, cancellationToken);
+                        workerTasks[i] = Task.Run(async() => await ProcessCompilationEventsCoreAsync(analysisScope, analysisStateOpt, prePopulatedEventQueue, cancellationToken).ConfigureAwait(false));
```
Comment (Pilchie) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Why?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 8]
```diff
@@ -845,35 +861,66 @@ private static Diagnostic GetFilteredDiagnostic(Diagnostic diagnostic, Compilati
             return compilation.Options.FilterDiagnostic(diagnostic);
         }
 
-        private static Task<AnalyzerActions> GetAnalyzerActionsAsync(
+        private static Task<Tuple<AnalyzerActions, ImmutableDictionary<DiagnosticAnalyzer, SemaphoreSlim>>> GetAnalyzerActionsAsync(
```
Comment (Pilchie) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Ugh.  `Tuple` is such an unreadable type.  How would any caller know what to do with this.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs [hunk 9]
```diff
@@ -132,20 +132,27 @@ internal partial class AnalyzerManager
...
         /// <summary>
-        /// Get all the analyzer actions to execute for the given analyzer against a given compilation.
+        /// Get tuple with:
+        /// (a) All the analyzer actions to execute for the given analyzer against a given compilation and
+        /// (b) A flag indicating if these actions can be executed concurrently.
         /// The returned actions include the actions registered during <see cref="DiagnosticAnalyzer.Initialize(AnalysisContext)"/> method as well as
         /// the actions registered during <see cref="CompilationStartAnalyzerAction"/> for the given compilation.
         /// </summary>
-        public async Task<AnalyzerActions> GetAnalyzerActionsAsync(DiagnosticAnalyzer analyzer, AnalyzerExecutor analyzerExecutor)
+        public async Task<Tuple<AnalyzerActions, bool>> GetAnalyzerActionsAsync(DiagnosticAnalyzer analyzer, AnalyzerExecutor analyzerExecutor)
```
Comment (Pilchie) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs]: Again, please just make a type for this.  Tuples make methods impossible.

Comment (Pilchie) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs]: In this case, is there a problem with just splitting it into two calls?  Looks like you just need the `sessionScope` to get the bool.  Is calling `GetSessionAnalysisScopeAsync` twice a problem?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 3 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Sure, will change.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 4 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Sure.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs (hunk 5 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs]: Actually, this is just a defensive check - analyzer driver spawn's multiple worker threads to process compilation events only if there machine has more than one core (see [here](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalyzerDriver.cs,60)). ConcurrencyAnalyzer used by this test needs multiple worker threads - first one just blocks till remaining callbacks are finished. If driver takes a lock for calling into ConcurrencyAnalyzer (which would be a bug), then this test will deadlock.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs (hunk 6 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs]: Sure.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 7 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: `ProcessCompilationEventsCoreAsync` will callback into the analyzer on the same thread, and if that never returns we will never spawn other workers. This change ensures that all callbacks happen on background tasks and we always create all workers upfront.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 8 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: I agree, but I wasn't sure a private method with a single callsite should introduce a new type wrapping the values.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs (hunk 9 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs]: Let me clean this up a bit and update the PR.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 2 above)
Comment (Pilchie) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: I'm +1 to @JohnHamby's point about Register versus Enable.  I like EnableConcurrentExecution.

Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Agreed.

**Comment by mavasani** [comment]
@Pilchie @JohnHamby I have addressed your feedback and updated the PR.

**Comment by Pilchie** [comment]
:+1:

**Comment by JohnHamby** [comment]
My "related actions" proposal turns out to be too pessimistic for some very important cases. There will be a follow-on discussion about the best way to deal with throttling concurrency in the presence of mutable state.

I sign off.

**Comment by mavasani** [comment]
Thanks John.

**Comment by mavasani** [comment]
I have filed https://github.com/dotnet/roslyn/issues/7228 to track John's proposal for enhancing the API.

**Pull request merged**